### PR TITLE
Some bulk refactorings

### PR DESCRIPTION
**Pull request by oreissig**
I went through the codebase of the root project and did the following refactorings:
- make private methods static when they are plain functions
- add @Deprecated when javadoc says the method is @deprecated
- remove unused code
- avoid synthetic accessors by making private fields/methods package-private
- make private fields final where possible
- fix some inconsistent whitespaces

None of this should change the behaviour in any way.

### CODE REVIEW

**Comment by oreissig** [comment]
Rebased from master due to conflicts.

**Comment by PascalSchumacher** [comment]
Thanks for the pull request. :)

I already cherry-picked a0aed6c6e523d176ebd480075d54aa9a3e19b472 which fixes some mistakes.

I support merging the rest of the commits, although I'm not sure how much a0aed6c6e523d176ebd480075d54aa9a3e19b472 or a032ed64a21434f12d2d6a560d55c7efb79eb986 add. With 0a78058ebc415e41b9523b77a42a92b907f169f1 there is a small change that somebody calls the methods from dynamic groovy, but of course it is important to remove unused stuff.

I'm thankful for opinions of other commiters on this.

**Comment by oreissig** [comment]
This one particular `getSize()` method in 0a78058 I actually deleted initially, but added it again when I went through the diffs and noticed the comment above. I decided to make the method @Deprecated as a neat trick to indicate in an IDE that this should not be called if written by accident. If you don't like this trick, I'll just leave the method as is.

**Comment by jwagenleitner** [comment]
For commit a032ed6 (avoid synthetic accessors by making private fields/methods package-private), I think removing `private` from members of private nested classes where the member is accessed by the top-level class would be ok.  However, I don't think members of top-level public classes should be changed from `private` to `package-private`.  If they weren't designed to be used by other classes in the package then they shouldn't be made visible just to avoid a synthetic method.  So overall I'd be -1 for this particular commit.

The other commits look good to me.

**Comment by oreissig** [comment]
I also thought about the widening of scope, that may be unintended, but if there's a synthetic accessor method generated, it actually is made available inside the package anyway? Scope-wise there's not much of a difference between a package-private field and a package-private getter, or is there one?

**Comment by jwagenleitner** [comment]
> but if there's a synthetic accessor method generated, it actually is made available inside the package anyway? 

To my knowledge the synthetic accessors are inserted by the Java compiler so that top-level and inner classes can access private members of each other.  From a coding perspective it is not really feasible for another class in the package to use these compiler generated accessors (i.e., they wont show up in IDE code completion).  By widening the scope of the members of the enclosing class encapsulation is being broken in order to avoid having a few synthetic accessors; the only benefit might be slight performance increase and that is even very questionable.  But many people work on the codebase and encapsulation is more important in my opinion in this context.  If someone working in the code see they can access the member they may think it was designed to be accessed by classes in the package when it actually was not.

> Scope-wise there's not much of a difference between a package-private field and a package-private getter, or is there one?

No I don't believe so.  In my original comment I was referring to synthetic accessors in general and not just those generated for methods.  I think again it comes back to encapsulation and breaking it just to avoid the synthetic accessors is not a good thing.

**Comment by oreissig** [comment]
Good point about IDEs respecting the synthetic flag, thanks! So I'll revert all the scoping changes in non-inner classes. For private inner classes the scope should not matter that much, since they as a whole are not accessible to any other classes, therefore I leave my changes there.

**Comment by oreissig** [comment]
Rebased from trunk due to conflicts.

The only cases left where I changed from private to package-private is for private inner classes now.

**Comment by jwagenleitner** [comment]
Because many of the commits touch the same files it is difficult to review just the changes for the synthetic accessors/package-private.  I think for bulk changes like this, even if minor, separate PR's would be easier to review and manage.

**Comment by PascalSchumacher** [comment]
I agree in the future I would prefer separate PR's.

Can you please squash c7fec364237ff496c88d7018b436254bb5b26945 and fcbe33e2e9abcaaed0d0c7b273ff810f5acbb9e6? Thanks!

**Comment by oreissig** [comment]
Yes I already realized this was a bad idea. I would've love to just fixup the initial scope changing commit, but could not do so without major effort because it conflicts with the other commit adding final here and there, that has been done in the meantime :-/

**Comment by PascalSchumacher** [comment]
Another problem with doing so many changes is that there are a lot of merge conflicts when I try to merge the changes into the 2_4_X branch, but without merging it's highly likely that there will be conflicts when backporting bugfixes. :(

**Comment by oreissig** [comment]
If you want I can go through again and restructure the changes. How do you like it? Individual PRs for each of the commits (except the fixup)? Should those individual PRs only contain one commit each, or do you like it to be split up by some metric?

**Comment by PascalSchumacher** [comment]
No further thinking about it I'm not sure if multiple pull request will really help in this case. I guess the pull request would have massive merge conflicts with each other.

I think 69ade9eb71829395e798fd8ba6395287bc44940a it the most valuable commit, because removing code dead code is good for readability and maintainability. I do not know if there is an easy way to untangle it from the rest? c3ad8298bd817e43fd0f56cb898fbce647002298 is also nice. Of course the other commit also have value.

**Comment by jwagenleitner** [comment]
I think that bulk refactorings like these are best done one at a time (serially) with each as it's own PR with a single commit focused on the one change and agree that the unused code commit would be a good one to start with.  For me I'd wait for it to be merged before starting work on the next PR if lots of the same files are to be touched, and work on them in order of most likely to least likely to be merged.  I think that would avoid the headaches of dealing with merge conflicts.

**Comment by PascalSchumacher** [comment]
@oreissig I rebased the remove unused code commit. Thanks!

File: src/main/groovy/lang/GroovyCodeSource.java [hunk 1]
```diff
@@ -164,7 +164,7 @@ public GroovyCodeSource(URI uri) throws IOException {
         this(uri.toURL());
     }
 
-    public GroovyCodeSource(URL url) throws IOException {
```
Comment (paulk-asert) [inline src/main/groovy/lang/GroovyCodeSource.java]: I only just spotted this now. While "unused" in the sense that this method didn't throw this exception anymore (I presume it did in the past), removing the exception breaks contract/source code compatibility. :-( I know it is a result of how checked exceptions are handled on the JVM but we should have not done this in a point release - all good to know in hindsight. Anyway, it isn't clear that reverting after a year is going to help things so I'm thinking of just documenting for now with a Jira issue. Apologies if this has already been raised elsewhere.

Comment (oreissig) [inline src/main/groovy/lang/GroovyCodeSource.java]: d'oh, I totally forgot about that :(

File: src/main/groovy/grape/GrapeIvy.groovy [hunk 2]
```diff
@@ -76,11 +76,11 @@ class GrapeIvy implements GrapeEngine {
...
     Set<String> downloadedArtifacts
     // weak hash map so we don't leak loaders directly
-    Map<ClassLoader, Set<IvyGrabRecord>> loadedDeps = new WeakHashMap<ClassLoader, Set<IvyGrabRecord>>()
+    final Map<ClassLoader, Set<IvyGrabRecord>> loadedDeps = new WeakHashMap<ClassLoader, Set<IvyGrabRecord>>()
     // set that stores the IvyGrabRecord(s) for all the dependencies in each grab() call
-    Set<IvyGrabRecord> grabRecordsForCurrDependencies = new LinkedHashSet<IvyGrabRecord>()
+    final Set<IvyGrabRecord> grabRecordsForCurrDependencies = new LinkedHashSet<IvyGrabRecord>()
     // we keep the settings so that addResolver can add to the resolver chain
-    IvySettings settings
+    final IvySettings settings
```
Comment (blackdrag) [inline src/main/groovy/grape/GrapeIvy.groovy]: not sure those finals are ok, it means there will be no getter for these anymore. So I would like to hear from for example Paul that these are not supposed to have getter

File: src/main/groovy/lang/ObjectRange.java [hunk 3]
```diff
@@ -270,7 +270,8 @@ protected int compareTo(Comparable first, Comparable second) {
     /**
      * protection against calls from Groovy
      */
-    private void setSize(int size) {
+    @Deprecated
+    void setSize(int size) {
         throw new UnsupportedOperationException("size must not be changed");
```
Comment (blackdrag) [inline src/main/groovy/lang/ObjectRange.java]: why is it deprecated again?

File: src/main/groovy/lang/ProxyMetaClass.java [hunk 4]
```diff
@@ -44,7 +42,7 @@
     /**
      * convenience factory method for the most usual case.
      */
-    public static ProxyMetaClass getInstance(Class theClass) throws IntrospectionException {
+    public static ProxyMetaClass getInstance(Class theClass) {
```
Comment (blackdrag) [inline src/main/groovy/lang/ProxyMetaClass.java]: removal of checked exception in public method

File: src/main/groovy/lang/ProxyMetaClass.java [hunk 5]
```diff
@@ -53,7 +51,7 @@ public static ProxyMetaClass getInstance(Class theClass) throws IntrospectionExc
     /**
      * @param adaptee the MetaClass to decorate with interceptability
      */
-    public ProxyMetaClass(MetaClassRegistry registry, Class theClass, MetaClass adaptee) throws IntrospectionException {
```
Comment (blackdrag) [inline src/main/groovy/lang/ProxyMetaClass.java]: removal of checked exception in public method

File: src/main/org/codehaus/groovy/reflection/GeneratedMetaMethod.java [hunk 6]
```diff
@@ -162,7 +162,7 @@ public static void saveDgmInfo(List<DgmMethodRecord> records, String file) throw
             out.close();
         }
 
-        public static List<DgmMethodRecord> loadDgmInfo() throws IOException, ClassNotFoundException {
+        public static List<DgmMethodRecord> loadDgmInfo() throws IOException {
```
Comment (blackdrag) [inline src/main/org/codehaus/groovy/reflection/GeneratedMetaMethod.java]: removal of checked exception from public api

File: src/main/org/codehaus/groovy/tools/DgmConverter.java [hunk 7]
```diff
@@ -37,7 +37,7 @@
 
 public class DgmConverter implements Opcodes {
 
-    public static void main(String[] args) throws IOException, ClassNotFoundException {
+    public static void main(String[] args) throws IOException {
```
Comment (blackdrag) [inline src/main/org/codehaus/groovy/tools/DgmConverter.java]: even though this is not likely to cause a problem, still removal of checked exception on public method

File: src/main/org/codehaus/groovy/tools/shell/IO.java [hunk 8]
```diff
@@ -130,7 +130,7 @@ public boolean isDebug() {
     /**
      * Flush both output streams.
      */
-    public void flush() throws IOException {
+    public void flush() {
```
Comment (blackdrag) [inline src/main/org/codehaus/groovy/tools/shell/IO.java]: removal of checked exception on public method

File: src/main/org/codehaus/groovy/vmplugin/v7/IndyArrayAccess.java [hunk 9]
```diff
@@ -96,29 +96,6 @@ private static MethodHandle buildSetter(Class arrayClass){
...
-        if (array instanceof char[]) return ((char[])array).length;
-        if (array instanceof short[]) return ((short[])array).length;
-        if (array instanceof int[]) return ((int[])array).length;
-        if (array instanceof long[]) return ((long[])array).length;
-        if (array instanceof float[]) return ((float[])array).length;
-        if (array instanceof double[]) return ((double[])array).length;
-        return 0;
-    }
-
-    private static int normalizeIndex(Object array, int i) {
```
Comment (blackdrag) [inline src/main/org/codehaus/groovy/vmplugin/v7/IndyArrayAccess.java]: normalizeIndex is used via MethodHandle in line 39. you cannot remove these

**Pull request closed without merging**